# CLASSIFYING AMAZON FOOD REVIEWS USING SUPPORT VECTOR MACHINES

In [1]:
# Importing the libraries required for our task

import sqlite3
import pandas as pd
import numpy as np
import nltk
import string
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import roc_curve, auc
from nltk.stem.porter import PorterStemmer

In [2]:
# Importing the required sqlite database which contains the reviews

con = sqlite3.connect('database.sqlite')

'''
Our objective is to check whether the review is positive or negative.

The dataset originally consists of reviews from 1 to 5. We will consider reviews which are rated as 4 and 5 to be positive, reviews 
which are rated 1 and 2 to be negative. As, we cannot draw any conclusions from review which is rated 3 star we will eliminate
all the reviews rated 3 star.

'''

# Filtering out the data w/o the 3 star reviews

filtered_data = pd.read_sql_query("""SELECT * FROM Reviews WHERE Score != 3""", con) 

In [3]:
# Checking whether the data has been filtered properly

filtered_data['Score'].unique()

# We can see that there is no 3 star review in the data

array([5, 1, 4, 2], dtype=int64)

In [4]:
# As we have eliminated the data with 3 start reviews, we will label the remaining data(4 and 5 scores) as positive and negative(1 and 2 scores).

# Creating a function to label the data
def partition(x):
    if x < 3:
        return '0'
    return '1'


# Applying the labels to the data

actualScore = filtered_data['Score']
positiveNegative = actualScore.map(partition) 
filtered_data['Score'] = positiveNegative

In [5]:
# Checking the first few rows of the data

filtered_data.head(10)

# We can see that the score has been changed to positive and negative instead of 5,4,1,2

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   
5   6  B006K2ZZ7K   ADT0SRK1MGOEU                   Twoapennything   
6   7  B006K2ZZ7K  A1SP2KVKFXXRU1                David C. Sullivan   
7   8  B006K2ZZ7K  A3JRGQVEQN31IQ               Pamela G. Williams   
8   9  B000E7L2R4  A1MZYO9TZK0BBI                         R. James   
9  10  B00171APVA  A21BT40VZCCYT4                    Carol A. Reed   

   HelpfulnessNumerator  HelpfulnessDenominator Score        Time  \
0                     1                       1     1  1303862400   
1                     0                       0     0  1346976000   
2                     1                       1     1  1219017600   
3                     3                       3     0  1307923200   
4                     0                       0     1  1350777600   
5                     0                       0     1  1342051200   
6                     0                       0     1  1340150400   
7                     0                       0     1  1336003200   
8                     1                       1     1  1322006400   
9                     0                       0     1  1351209600   

                                         Summary  \
0                          Good Quality Dog Food   
1                              Not as Advertised   
2                          "Delight" says it all   
3                                 Cough Medicine   
4                                    Great taffy   
5                                     Nice Taffy   
6  Great!  Just as good as the expensive brands!   
7                         Wonderful, tasty taffy   
8                                     Yay Barley   
9                               Healthy Dog Food   

                                                Text  
0  I have bought several of the Vitality canned d...  
1  Product arrived labeled as Jumbo Salted Peanut...  
2  This is a confection that has been around a fe...  
3  If you are looking for the secret ingredient i...  
4  Great taffy at a great price.  There was a wid...  
5  I got a wild hair for taffy and ordered this f...  
6  This saltwater taffy had great flavors and was...  
7  This taffy is so good.  It is very soft and ch...  
8  Right now I'm mostly just sprouting this so my...  
9  This is a very healthy dog food. Good for thei...

In [6]:
# Let us look at the shape of the data

filtered_data.shape

(525814, 10)

In [7]:
# Dropping any duplicates if they are present in the data

duplicates_dropped=filtered_data.drop_duplicates(subset={"UserId","ProfileName","Time","Text"}, keep='first', inplace=False)
duplicates_dropped.shape

(364173, 10)

In [8]:
# Eliminating the rows where helpfulness numerator is greator than the helfulness denominator

final=duplicates_dropped[duplicates_dropped.HelpfulnessNumerator<=duplicates_dropped.HelpfulnessDenominator]

In [9]:
# Taking a look at the shape of the data

final.shape

(364171, 10)

In [10]:
# We will select only the required columns of which we will filter some reviews and assign them their labels and sort them wrt time

final_data = final[['ProductId','Time','Text','Score']]

In [11]:
# Checking the dimension of the data
final_data.shape


(364171, 4)

In [12]:
# Randomly selecting some reviews from the 'final_data' data

# First we will extract the values from the given dataframe

X=final_data.iloc[:,:].values

# randomly extracting 15000 reviews from the dataset

import random

n = 364171
m = 15000
p = m/n

sampled_data = [];

for i in range(0,n):
    if random.random() <= p:
        sampled_data.append(X[i,:])

In [13]:
# Assigning the extracted data to a dataframe

names = ['ProductId','Time','Text','Score']

sample = pd.DataFrame(sampled_data,columns= names)

In [14]:
# Checking the dimensions of the sampled data

sample.shape

(14976, 4)

In [15]:
# Checking the first few rows of the data

sample.head()

ProductId        Time                                               Text  \
0  B001GVISJM  1324598400  Twizzlers, Strawberry my childhood favorite ca...   
1  B001EO5QW8  1255392000  McCann's Oatmeal is a good quality choice.  Ou...   
2  B005R8JE8O  1344902400  i know i cannot make tea this good.  granted, ...   
3  B0019CW0HE  1335398400  I have a standard poodle and pomeranian who bo...   
4  B0037LW78C  1337990400  I have been drinking Royal King 100% Natural O...   

  Score  
0     1  
1     1  
2     1  
3     1  
4     1

In [16]:
# Now we will sort the data according to timestamp

sorted_data=sample.sort_values('Time', axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last')

In [17]:
# Checking if the data has been sorted or not

sorted_data.head(15)

ProductId        Time  \
6924   B00004RYGX   959990400   
4785   B00004S1C6  1035504000   
8039   B0000TU8VM  1068249600   
4834   B0000DHZY1  1068508800   
11221  B0000D17W5  1070323200   
10494  B0000DBN1H  1072656000   
11861  B0000CA4TK  1073865600   
149    B00016UX0K  1081555200   
13801  B0000DG87B  1082073600   
6607   B00005IX98  1082678400   
11747  B000084E76  1083024000   
4733   B0000TA3QM  1083024000   
6923   B00004RYGX  1089504000   
51     B000WNJ73Q  1091318400   
13309  B0002PPW56  1093305600   

                                                    Text Score  
6924   I'm getting crazy.I'm looking for Beatlejuice ...     1  
4785   The spectra paste food colors are simply fanta...     1  
8039   A piquant, sweet, colorful treat.  Don't be pu...     1  
4834   I have baked with this organic vanilla in the ...     0  
11221  This stuff is perfect for a camping trip or wh...     1  
10494  For a bag tea--which generally are not as good...     1  
11861  I also live in exile in San Francisco, but I'm...     1  
149    Mae Ploy Sweet Chili Sauce is becoming a stand...     1  
13801  The tree came in great condition, healthy and ...     1  
6607   I can't say enough good things about the Espre...     1  
11747  My cat Foutchie is heading toward 5 now and no...     1  
4733   I bought the Raisin Brown Bread today on a whi...     1  
6923   a couple dies.they live in a house as spirits ...     1  
51     Five of my five dogs agree - they'd rather mun...     1  
13309  This is a serious espresso. 100% Pure Espresso...     1

In [18]:
# We can see that the data has been sorted wrt time, now we will only consider the 'Text' and 'Score' columns henceforth

sorted_final = sorted_data[['Text','Score']]


In [19]:
# Checking the first few rows of the data to ensure we have the right data

sorted_final.head()

Text Score
6924   I'm getting crazy.I'm looking for Beatlejuice ...     1
4785   The spectra paste food colors are simply fanta...     1
8039   A piquant, sweet, colorful treat.  Don't be pu...     1
4834   I have baked with this organic vanilla in the ...     0
11221  This stuff is perfect for a camping trip or wh...     1

In [20]:
# The next task is to clean the text data so that it can be fed to the model

# Checking if there are unknown elements in the data

# find sentences containing HTML tags

import re

i=0;
for sent in sorted_final['Text'].values:
    if (len(re.findall('<.*?>', sent))):
        print(i)
        print(sent)
        break;
    i += 1;   
    
# We can see that the data contains html tags, we will need to remove those

0
I'm getting crazy.I'm looking for Beatlejuice french version video.<p>Is it really impossible today not to find the French VHS version of this film ?<p>Could U please tell me something about it ? Tks


In [21]:
'''
We will perform the data cleaning steps on the text data.
For that we will import some packages for stopwords removal, word stemmatization and cleaning html and punctuation marks.
'''
import re
from nltk.corpus import stopwords

nltk.download('stopwords')

stop = set(stopwords.words('english')) #set of stopwords
sno = nltk.stem.SnowballStemmer('english') #initialising the snowball stemmer

def cleanhtml(sentence): #function to clean the word of any html-tags
    cleanr = re.compile('<.*?>')
    cleantext = re.sub(cleanr, ' ', sentence)
    return cleantext
def cleanpunc(sentence): #function to clean the word of any punctuation or special characters
    cleaned = re.sub(r'[?|!|\'|"|#]',r'',sentence)
    cleaned = re.sub(r'[.|,|)|(|\|/]',r' ',cleaned)
    return  cleaned

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\kulkarni\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [22]:
#Code for implementing step-by-step the checks mentioned in the pre-processing phase

i=0
str1=' '
final_string=[]
all_positive_words=[] # store words from +ve reviews here
all_negative_words=[] # store words from -ve reviews here.
s=''
for sent in sorted_final['Text'].values:
    filtered_sentence=[]
    #print(sent);
    sent=cleanhtml(sent) # remove HTMl tags
    for w in sent.split():
        for cleaned_words in cleanpunc(w).split():
            if((cleaned_words.isalpha()) & (len(cleaned_words)>2)):    
                if(cleaned_words.lower() not in stop):
                    s=(sno.stem(cleaned_words.lower())).encode('utf8')
                    filtered_sentence.append(s)
                    if (final['Score'].values)[i] == 'positive': 
                        all_positive_words.append(s) #list of all words used to describe positive reviews
                    if(final['Score'].values)[i] == 'negative':
                        all_negative_words.append(s) #list of all words used to describe negative reviews reviews
                else:
                    continue
            else:
                continue 
    #print(filtered_sentence)
    str1 = b" ".join(filtered_sentence) #final string of cleaned words
    #print("***********************************************************************")
    
    final_string.append(str1)
    i+=1

In [23]:
sorted_final['Cleaned_Text'] = final_string

C:\Users\kulkarni\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


# We will do the classification based on various representations of the text data i.e BoW, Tf-Idf

# Using the Bag of Words text representation to create our first Naive Bayes models

In [24]:
# Creating the BoW representation of the text

count_vect = CountVectorizer() #in scikit-learn
final_counts = count_vect.fit_transform(sorted_final['Cleaned_Text'].values).todense()

In [25]:
# Assigning the final_tf_idf data to 'X' variable

X = final_counts

In [26]:
# Assinging the score to 'y' variable

y = np.array(sorted_final['Score'])

# We will first split the data into train and test sets, then split the train data in to train and cross-validation sets

In [27]:
# Splitting the data into train test sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3) # Splitting train test with 70:30 ratio


In [28]:
# Splitting the train data into cross validation and test datasets

X_tr, X_CV, y_tr, y_CV = train_test_split(X_train, y_train, test_size=0.3) # Splitting train cross-val with 70:30 ratio

# We will be using the SGDClassifier algorithm to classify the data as positive and negative. Here, we will use  'alpha' as the hyperparameter which will take various values to determine the best classifier depending upon the hyperparameter.

# We will detemine the best value of the hyperparameter based on the cross validation accuracy

In [29]:
from sklearn import linear_model

In [30]:
# Importing the 

from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score

# We have taken the following values for 'alpha'
alpha_values = [0.00001, 0.0001, 0.001, 0.1, 1, 10, 100,1000]

for i in alpha_values:
    clf = linear_model.SGDClassifier(alpha = i)
    clf.fit(X_tr, y_tr)
    pred = clf.predict(X_CV)
    
    # evaluate CV accuracy
    acc = accuracy_score(y_CV, pred, normalize=True) * float(100)
    print('\nCV accuracy for alpha = %.5f is %d%%' % (i, acc))

C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)



CV accuracy for alpha = 0.00001 is 87%

CV accuracy for alpha = 0.00010 is 87%

CV accuracy for alpha = 0.00100 is 89%

CV accuracy for alpha = 0.10000 is 83%

CV accuracy for alpha = 1.00000 is 83%

CV accuracy for alpha = 10.00000 is 83%

CV accuracy for alpha = 100.00000 is 83%

CV accuracy for alpha = 1000.00000 is 16%


# We can see that for the alpha value of 0.001 we get the maximum cross validation accuracy.

# Creating a SGDClassifier model with BoW text representation with an alpha value of 0.001

In [31]:
sgd = linear_model.SGDClassifier(alpha = 0.001)
sgd.fit(X_tr, y_tr)
pred_acc = sgd.predict(X_test)
acc = accuracy_score(y_test, pred_acc, normalize=True) * float(100)
print('\nTest accuracy for alpha = 0.001 is %d%%' % (acc))

C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)



Test accuracy for alpha = 0.001 is 88%


# Using the TF-IDF text representation to create our Naive Bayes models

In [32]:
# In this stage we have initialized the tf-idf vectorizer and applied it to the text data which has been stored in the final_tf_idf vriable

tf_idf_vect = TfidfVectorizer()
final_tf_idf = tf_idf_vect.fit_transform(sorted_final['Cleaned_Text'].values).todense()

In [33]:
# Assigning the final_tf_idf data to 'X1' variable

X1 = final_tf_idf

In [34]:
# Assinging the score to 'y1' variable

y1 = np.array(sorted_final['Score'])

In [35]:
# Splitting the data into train test sets

from sklearn.model_selection import train_test_split

X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y1, test_size=0.3) # Splitting train test with 70:30 ratio

In [38]:
# Splitting the train data into train and cross validation sets

X_tr1, X_CV1, y_tr1, y_CV1 = train_test_split(X_train1, y_train1, test_size=0.3) # Splitting train cross-val with 70:30 ratio

In [38]:
# We have taken the following values for 'alpha'
alpha_values = [0.00001, 0.0001, 0.001, 0.1, 1, 10, 100,1000]

for i in alpha_values:
    clf = linear_model.SGDClassifier(alpha = i)
    clf.fit(X_tr1, y_tr1)
    pred1 = clf.predict(X_CV1)
    
    # evaluate CV accuracy
    acc1 = accuracy_score(y_CV1, pred1, normalize=True) * float(100)
    print('\nCV accuracy for alpha = %.5f is %d%%' % (i, acc1))

C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)



CV accuracy for alpha = 0.00001 is 88%

CV accuracy for alpha = 0.00010 is 89%

CV accuracy for alpha = 0.00100 is 83%

CV accuracy for alpha = 0.10000 is 83%

CV accuracy for alpha = 1.00000 is 83%

CV accuracy for alpha = 10.00000 is 83%

CV accuracy for alpha = 100.00000 is 83%

CV accuracy for alpha = 1000.00000 is 83%


# We can see that for the alpha value of 0.001 we get the maximum cross validation accuracy.

# Creating a SGDClassifier model with TF-IDF text representation with an alpha value of 0.001

In [40]:
sgd_tfidf = linear_model.SGDClassifier(alpha = 0.0001)
sgd_tfidf.fit(X_tr1, y_tr1)
pred_acc1 = sgd_tfidf.predict(X_test1)
acc_tfidf = accuracy_score(y_test1, pred_acc1, normalize=True) * float(100)
print('\nTest accuracy for alpha = 0.001 is %d%%' % (acc_tfidf))

C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)



Test accuracy for alpha = 0.001 is 89%


# Using the W2Vec representation of the text data to apply the SGD Classifier

In [41]:
# Importing the required models for the project

from gensim.models import Word2Vec
from gensim.models import KeyedVectors
import pickle

C:\Users\kulkarni\Anaconda3\lib\site-packages\gensim\utils.py:1197: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")


In [42]:
import gensim
i=0
list_of_sent=[]
for sent in sorted_final['Text'].values:
    filtered_sentence=[]
    sent=cleanhtml(sent)
    for w in sent.split():
        for cleaned_words in cleanpunc(w).split():
            if(cleaned_words.isalpha()):    
                filtered_sentence.append(cleaned_words.lower())
            else:
                continue 
    list_of_sent.append(filtered_sentence)
    

In [43]:
w2v_model=gensim.models.Word2Vec(list_of_sent,min_count=5,size=50, workers=3)

In [44]:
words = list(w2v_model.wv.vocab)
print(len(words))

7752


# Creating an Avg W2Vec representation of each review

In [45]:
# Computing the Avg W2Vec representation of each review and storing it in 'sent_vectors' list

sent_vectors = []; # the avg-w2v for each sentence/review is stored in this list
for sent in list_of_sent: # for each review/sentence
    sent_vec = np.zeros(50) # as word vectors are of zero length
    cnt_words =0; # num of words with a valid vector in the sentence/review
    for word in sent: # for each word in a review/sentence
        try:
            vec = w2v_model.wv[word]
            sent_vec += vec
            cnt_words += 1
        except:
            pass
    sent_vec /= cnt_words
    sent_vectors.append(sent_vec)
print(len(sent_vectors))
print(len(sent_vectors[0]))

14976
50


In [46]:
# Storing the list in a variable 'X2'

X2 = sent_vectors

In [47]:
# Storing the review score in a 'y2' variable

y2 = sorted_final['Score']

In [55]:
# Splitting the data into train test sets

#from sklearn.model_selection import train_test_split

X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size=0.3) # Splitting train test with 70:30 ratio

In [56]:
# Splitting the train data into train and cross validation sets

X_tr2, X_CV2, y_tr2, y_CV2 = train_test_split(X_train2, y_train2, test_size=0.3) # Splitting train cross-val with 70:30 ratio

In [58]:
# We have taken the following values for 'alpha'
alpha_values = [0.00001, 0.0001, 0.001, 0.1, 1, 10, 100,1000]

for i in alpha_values:
    clf_avgw2vec = linear_model.SGDClassifier(alpha = i)
    clf_avgw2vec.fit(X_tr2, y_tr2)
    pred2 = clf_avgw2vec.predict(X_CV2)
    
    # evaluate CV accuracy
    acc2 = accuracy_score(y_CV2, pred2, normalize=True) * float(100)
    print('\nCV accuracy for alpha = %.5f is %d%%' % (i, acc2))

C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)



CV accuracy for alpha = 0.00001 is 79%

CV accuracy for alpha = 0.00010 is 85%

CV accuracy for alpha = 0.00100 is 85%

CV accuracy for alpha = 0.10000 is 84%

CV accuracy for alpha = 1.00000 is 84%

CV accuracy for alpha = 10.00000 is 84%

CV accuracy for alpha = 100.00000 is 84%

CV accuracy for alpha = 1000.00000 is 84%


# We can see that for the alpha value of 0.0001 we get the maximum cross validation accuracy.

# Creating a SGDClassifier model with avgW2Vec text representation with an alpha value of 0.0001

In [61]:
sgd_avgw2vec = linear_model.SGDClassifier(alpha = 0.0001)
sgd_avgw2vec.fit(X_tr2, y_tr2)
pred_acc2 = sgd_avgw2vec.predict(X_test2)
acc_avgw2vec = accuracy_score(y_test2, pred_acc2, normalize=True) * float(100)
print('\nTest accuracy for alpha = 0.0001 is %d%%' % (acc_avgw2vec))


Test accuracy for alpha = 0.0001 is 84%


C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


# Creating a Tf-Idf weighted W2Vec representation of each review

In [62]:
# TF-IDF weighted Word2Vec
tfidf_feat = tf_idf_vect.get_feature_names() # tfidf words/col-names
# final_tf_idf is the sparse matrix with row= sentence, col=word and cell_val = tfidf

tfidf_sent_vectors = []; # the tfidf-w2v for each sentence/review is stored in this list
row=0;
for sent in list_of_sent: # for each review/sentence
    sent_vec = np.zeros(50) # as word vectors are of zero length
    weight_sum =0; # num of words with a valid vector in the sentence/review
    for word in sent: # for each word in a review/sentence
        try:
            vec = w2v_model.wv[word]
            # obtain the tf_idfidf of a word in a sentence/review
            tf_idf = final_tf_idf[row, tfidf_feat.index(word)]
            #tfidf = final_tf_idf[row, tfidf_feat.index(word)]
            sent_vec += (vec * tf_idf)
            weight_sum += tf_idf
        except:
            pass
    sent_vec /= weight_sum
    tfidf_sent_vectors.append(sent_vec)
    row += 1

C:\Users\kulkarni\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: RuntimeWarning: invalid value encountered in true_divide


In [63]:
# Checking for any 'NaN' values in the data

np.isnan(tfidf_sent_vectors).any()

# We can see that there are na values in the data

True

In [64]:
# We will replace all the 'Nan' values w/ mean of the respective columns


from sklearn.preprocessing import Imputer
imp = Imputer(missing_values='NaN', strategy='mean', axis=0)
data_tfidf = imp.fit_transform(tfidf_sent_vectors) # Assigning the imputed matrix w/o Nan values to data variable

In [65]:
# Checking if our imputation has succeded

np.isnan(data_tfidf).any()

# We can see that there are no na values any more.

False

In [66]:
# Assigning the 'data_tfidf' matrix to 'X3' variable

X3 = data_tfidf

In [67]:
# Assigning the scores to the 'y3' variable

y3 = sorted_final['Score']

In [68]:
# Splitting the data into train test sets

#from sklearn.model_selection import train_test_split

X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y3, test_size=0.3) # Splitting train test with 70:30 ratio

In [69]:
# Splitting the train data into train and cross validation sets

X_tr3, X_CV3, y_tr3, y_CV3 = train_test_split(X_train3, y_train3, test_size=0.3) # Splitting train cross-val with 70:30 ratio

In [72]:
# We have taken the following values for 'alpha'
alpha_values = [0.00001, 0.0001, 0.001, 0.1, 1, 10, 100,1000]

for i in alpha_values:
    clf_tfidfw2vec = linear_model.SGDClassifier(alpha = i)
    clf_tfidfw2vec.fit(X_tr3, y_tr3)
    pred3 = clf_tfidfw2vec.predict(X_CV3)
    
    # evaluate CV accuracy
    acc3 = accuracy_score(y_CV3, pred3, normalize=True) * float(100)
    print('\nCV accuracy for alpha = %.5f is %d%%' % (i, acc3))


CV accuracy for alpha = 0.00001 is 79%

CV accuracy for alpha = 0.00010 is 81%

CV accuracy for alpha = 0.00100 is 84%

CV accuracy for alpha = 0.10000 is 84%

CV accuracy for alpha = 1.00000 is 84%

CV accuracy for alpha = 10.00000 is 84%

CV accuracy for alpha = 100.00000 is 84%

CV accuracy for alpha = 1000.00000 is 84%


C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


In [73]:
sgd_tfidfw2vec = linear_model.SGDClassifier(alpha = 0.001)
sgd_tfidfw2vec.fit(X_tr3, y_tr3)
pred_acc3 = sgd_tfidfw2vec.predict(X_test3)
acc_tfidfw2vec = accuracy_score(y_test3, pred_acc3, normalize=True) * float(100)
print('\nTest accuracy for alpha = 0.001 is %d%%' % (acc_tfidfw2vec))


Test accuracy for alpha = 0.001 is 84%


C:\Users\kulkarni\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


# We have applied the SGD Classifier to the Amazon Food Reviews dataset using various text representations of the data and have achieved varying results.

# The accuracies we have acheived are as follows:

# In this case we have used only 15000 reviews off the 300000+ available due to memory constraints. Still we are getting high accuracies such as 89%. 

# When we will use more data the training will be much better and we can look forward to have much better results